### PR TITLE
[kafka-clusters] add support for multi_az and service accounts

### PR DESCRIPTION
**Pull request by maorfr**
part of https://issues.redhat.com/browse/APPSRE-2728

for a consumer to use a cluster, they need a service account. this PR adds support for creating a service account per cluster (with the cluster name being the service account name) and returns the credentials for that service account in the output Secret to be consumed by applications.

### CODE REVIEW

File: reconcile/kafka_clusters.py [hunk 1]
```diff
@@ -102,14 +119,15 @@ def run(dry_run, thread_pool_size=10,
             error = True
             continue
         # check if cluster is ready. if not - wait
-        if current_cluster['status'] != 'complete':
+        if current_cluster['status'] != 'ready':
```
Comment (Piojo) [inline reconcile/kafka_clusters.py:122]: Can we convert these magic strings into constants? A mistake on a future iteration, (say, I'm pushing this during an outage and I type `rready`) will waste hours of debugging.

File: reconcile/utils/ocm.py [hunk 2]
```diff
@@ -633,12 +633,35 @@ def get_kafka_clusters(self, fields=None):
...
+    def get_kafka_service_accounts(self, fields=None):
+        """Returns details of the Kafka service accounts """
+        results = []
+        api = '/api/managed-services-api/v1/serviceaccounts'
+        service_accounts = self._get_json(api)['items']
+        for sa in service_accounts:
+            sa_id = sa['id']
+            id_api = f'{api}/{sa_id}'
+            sa_details = self._get_json(id_api)
+            item = {k: v for k, v in sa_details.items() if k in fields}
```
Comment (Piojo) [inline reconcile/utils/ocm.py:645]: What happens if `fields` has its default value of `None`? Won't this crash?

File: reconcile/utils/ocm.py [hunk 3]
```diff
@@ -633,12 +633,35 @@ def get_kafka_clusters(self, fields=None):
...
+
     def create_kafka_cluster(self, data):
         """Creates (async) a Kafka cluster """
         api = '/api/managed-services-api/v1/kafkas'
         params = {'async': 'true'}
         self._post(api, data, params)
 
+    def create_kafka_service_account(self, name):
+        """ Creates a Kafka service account """
+        api = '/api/managed-services-api/v1/serviceaccounts'
```
Comment (Piojo) [inline reconcile/utils/ocm.py:657]: This looks like a constant we can share across the entire class, doesn't it?

File: reconcile/kafka_clusters.py [hunk 4]
```diff
@@ -75,10 +76,26 @@ def run(dry_run, thread_pool_size=10,
...
 
     error = False
     for kafka_cluster in kafka_clusters:
         kafka_cluster_name = kafka_cluster['name']
+        # get a service account for the cluster
+        # we match cluster to service account by name
+        service_accounts = [sa for sa in kafka_service_accounts
+                            if sa['name'] == kafka_cluster_name]
+        if service_accounts:
+            service_account = service_accounts[0]
```
Comment (Piojo) [inline reconcile/kafka_clusters.py:89]: Please, please, please, let's avoid variables that differ from just one character. As we get tired we'll confuse them.

In this context, `sa` or `acc` are perfectly valid identifiers.

File: reconcile/kafka_clusters.py (hunk 1 above)
Comment (maorfr) [reply reconcile/kafka_clusters.py]: done

File: reconcile/utils/ocm.py (hunk 2 above)
Comment (maorfr) [reply reconcile/utils/ocm.py]: good catch! fixed

File: reconcile/utils/ocm.py (hunk 3 above)
Comment (maorfr) [reply reconcile/utils/ocm.py]: i usually like copying the api directly from https://api.openshift.com, but we can go with constants.

File: reconcile/kafka_clusters.py (hunk 4 above)
Comment (maorfr) [reply reconcile/kafka_clusters.py]: ok!

**Comment by maorfr** [comment]
note: more changes are about to be added to handle storing the clientSecret in vault.

**Comment by maorfr** [comment]
image

**Comment by maorfr** [comment]
thanks for the review @Piojo! this is ready for another round.

**Comment by maorfr** [comment]
as I want this promoted before my week ends, i'm going to merge this. please feel free to provide a post-merge review and i promise to address everything. ask @rporres, i'm good for it.

**Pull request merged**